### PR TITLE
Python Jiterator supports multiple outputs

### PR DESCRIPTION
**Pull request by SherlockNoMad**
This PR is part3. 
Part1: https://github.com/pytorch/pytorch/pull/77902
Part2: https://github.com/pytorch/pytorch/pull/77921

Python Jiterator now supports returning multiple outputs

```
fn = torch.cuda.jiterator._create_multi_output_jit_fn(
"""
template <typename T>
T binary_2outputs(T i0, T i1, T& out0, T& out1) {
    out0 = i0 + i1;
    out1 = i0 - i1;
}
""",
num_outputs=2)

x = torch.rand(3, device='cuda')
y = torch.rand(3, device='cuda')
out0, out1 = fn(x, y) 

torch.allclose(out0, x+y)
torch.allclose(out1, x-y)
```

### CODE REVIEW

File: aten/src/ATen/cuda/jiterator.cu [hunk 1]
```diff
@@ -308,38 +311,47 @@ void jitted_gpu_kernel_dynamic(
...
-  jitted_gpu_kernel_dynamic_impl(kernel_name, iter, f, needs_dynamic_casting, extra_args);
+  jitted_gpu_kernel_dynamic_impl(kernel_name, iter, f, needs_dynamic_casting, extra_args, return_by_ref);
 }
 
 } // namespace native
 
 namespace cuda {
 
-at::Tensor CompileAndLaunchKernel(
+std::vector<at::Tensor> CompileAndLaunchKernel(
```
Comment (ngimel) [inline aten/src/ATen/cuda/jiterator.cu:321]: you probably want SmallVector here to reduce overheads (and for input tensors also)

**Review by ngimel — APPROVED** [review]
This is great!

File: aten/src/ATen/cuda/jiterator.cu (hunk 1 above)
Comment (SherlockNoMad) [reply aten/src/ATen/cuda/jiterator.cu]: done.

**Pull request closed without merging**